cura/Machines/MaterialManager.py

```diff
@@ -0,0 +1,324 @@
+from collections import defaultdict, OrderedDict
+from typing import Optional
+
+from PyQt5.Qt import QTimer, QObject, pyqtSignal
+
+from UM.Logger import Logger
+from UM.Settings import ContainerRegistry
+
+from cura.Machines.ContainerNode import ContainerNode
+
+
+class MaterialGroup:
+    __slots__ = ("name", "root_material_node", "derived_material_node_list")
+
+    def __init__(self, name: str):
+        self.name = name
+        self.root_material_node = None
+        self.derived_material_node_list = []
+
+    def __str__(self) -> str:
+        return "%s[%s]" % (self.__class__.__name__, self.name)
+
+
+class MaterialNode(ContainerNode):
+    __slots__ = ("material_map", "children_map")
+
+    def __init__(self, metadata: Optional[dict] = None):
+        super().__init__(metadata = metadata)
+        self.material_map = {}
+        self.children_map = {}
+
+
+class MaterialManager(QObject):
+
+    materialsUpdated = pyqtSignal()  # Emitted whenever the material lookup tables are updated.
+
+    def __init__(self, container_registry, parent = None):
+        super().__init__(parent)
+        self._container_registry = container_registry  # type: ContainerRegistry
+
+        self._fallback_materials_map = dict()  # material_type -> generic material metadata
+        self._material_group_map = dict()  # root_material_id -> MaterialGroup
+        self._diameter_machine_variant_material_map = dict()  # diameter -> dict(machine_definition_id -> MaterialNode)
```

Comment:
Consider making this a mapping from `(machine_definition_id, diameter)` to `MaterialNode`. As in, group the 2 keys together in a tuple. That would make more sense in how you're using this.

Reply:
We think using a simple `str` is easier to understand here and also it will be easier to make a view in GodMode to navigate through the trees.